### PR TITLE
system service: unset NOTIFY_SOCKET

### PR DESCRIPTION
**Pull request by vrothberg**
Unset the NOTIFY_SOCKET environment variable after sending the MAIN_PID and READY message.  This avoids any unintentional side-effects of other code paths using the socket assuming they'd run in a non-server short-lived Podman process.

#### Does this PR introduce a user-facing change?

```release-note
When running under systemd (e.g., via Quadlet), Podman will extend the start timeout in 30s steps up to 5min max. when pulling an image.
```

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
@edsantiago PTAL
To prevent a heart attack: I am not sure whether you like the GDB trick but I didn't find a more "straight-forward" solution to list the _current_ environment of another process. /proc/$pid/environ lists the env at process creation which is outdated.

File: test/system/260-sdnotify.bats [hunk 1]
```diff
@@ -503,4 +503,48 @@ none | false | false | 0
...
+
+    mainpid=${lines[0]:8}
+    assert "$(</proc/$mainpid/environ)" =~ ".*NOTIFY_SOCKET=$NOTIFY_SOCKET.*"
+    # /proc/$pid/environ includes the env when creating the process, so use GDB
+    # to print the current env variable.
+    #
+    # GDB trick found at https://unix.stackexchange.com/questions/29128/how-to-read-environment-variables-of-a-process
+    echo 'p (char *) getenv("FOO")' | gdb --quiet -p "$mainpid" | sed -n 's/.*"\(.*\)"$/\1/p'> $envfile
+    assert "$(<$envfile)" !~ ".*NOTIFY_SOCKET.*"
+
```
Comment (vrothberg) [inline test/system/260-sdnotify.bats:547]: Should kill $mainpid here.

**Comment by Luap99** [comment]
Requiring gdb means that will be a new dependency that everyone (gating, CI, local) needs to have installed. I don't think it is important to actually the test the env here.
Couldn't you just start a container with --sdnotify=container via the API and make sure the socket is not leaked into the container?  Because the service unset the env var so there is nothing to leak in such case and IMO that would be the important bug fix that this fixes.

**Comment by vrothberg** [comment]
> Couldn't you just start a container with --sdnotify=container via the API and make sure the socket is not leaked into the container?

That's less fun but definitely the better option, thanks!

**Comment by vrothberg** [comment]
> That's less fun but definitely the better option, thanks!

Turns out `--sdnotify=container` never worked on remote.  It doesn't make sense fore remote but I somehow expected it to break without this patch.

**Comment by edsantiago** [comment]
This is a toughie. FWIW I tried something like
* start server
* podman-remote run -d ... top
* podman inspect that container, get conmon pid
* tr '\0' '\012' /proc/conmonpid/env looking for NOTIFY_SOCKET

Nope, NOTIFY_SOCKET doesn't get to conmon, nor to the podman process either. (This is with regular podman, sans your patch. I wanted to see how to test this).

Got nowhere, and I need to give up for today. Will look again on Monday.

(Oh, and, yeah, let's try everything possible before resorting to gdb please)

**Comment by vrothberg** [comment]
@edsantiago let's wait for https://github.com/containers/common/pull/1616 to progress.  In case we move forward with https://github.com/containers/common/pull/1616, we can test unsetting NOTIFY_SOCKET more easily.  Namely, an `image pull` would not yield a specific DBUS message on the remote server.

**Comment by vrothberg** [comment]
@edsantiago WDYT of the new tests?

**Comment by edsantiago** [comment]
@vrothberg, sorry, what new tests? I don't see a new push?

**Comment by vrothberg** [comment]
The heat must be killing me at the moment - I was convinced of having pushed yesterday evening.  I'll push in a jiffy.

**Comment by vrothberg** [comment]
The jiffy got boiled as well :^)  New tests are up.

**Comment by edsantiago** [comment]
@vrothberg is it safe to assume that you're looking into and resolving the `USEC` failures?

**Comment by vrothberg** [comment]
> @vrothberg is it safe to assume that you're looking into and resolving the `USEC` failures?

Thanks!  Yes, I am on it.  Will push in $jiffy.

**Comment by vrothberg** [comment]
Should pass now.  Needed to make https://github.com/containers/common/pull/1616 a bit smarter as it was sending the USEC on --pull=newer even when the image was up to date.

File: test/system/260-sdnotify.bats [hunk 2]
```diff
@@ -503,4 +505,79 @@ none | false | false | 0
...
+    # inside a systemd unit (i.e., with NOTIFY_SOCKET set).
+    image_on_local_registry=localhost:${PODMAN_LOGIN_REGISTRY_PORT}/name:tag
+    registry_flags="--tls-verify=false --creds ${PODMAN_LOGIN_USER}:${PODMAN_LOGIN_PASS}"
+    start_registry
+
+    export NOTIFY_SOCKET=$PODMAN_TMPDIR/notify.sock
+    _start_socat
+
+    run_podman push $registry_flags $IMAGE $image_on_local_registry
+    run_podman pull $registry_flags $image_on_local_registry
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply test/system/260-sdnotify.bats:520]: @edsantiago before I repush, do you have other comments?

[the remark this answers is not in the document]
Comment (vrothberg) [reply test/system/260-sdnotify.bats:520]: `/bin/kill --timeout 5000 KILL --signal TERM $mainpid` instead?

[the remark this answers is not in the document]
Comment (edsantiago) [reply test/system/260-sdnotify.bats:520]: TIL about `kill --timeout`! That looks promising, and seems to work as I would hope (i.e. if the process exits, `kill` does not continue blocking). I think that's worth trying, but we need to make sure that fedora-old and debian all have the necessary version of `util-linux` ...

File: test/system/260-sdnotify.bats [hunk 3]
```diff
@@ -503,4 +505,79 @@ none | false | false | 0
...
+    run cat $_SOCAT_LOG
+    # The 'echo's help us debug failed runs
+    echo "socat log:"
+    echo "$output"
+    is "$output" "EXTEND_TIMEOUT_USEC=30000000"
```
Comment (edsantiago) [inline test/system/260-sdnotify.bats:525]: It has taken me this long just to understand this one line. Wow, this is tricky! LGTM, but wow.

File: test/system/260-sdnotify.bats [hunk 4]
```diff
@@ -503,4 +505,79 @@ none | false | false | 0
...
+
+    # Now pull remotely and make sure that the service does _not_ extend the
+    # timeout; the NOTIFY_SOCKET should be unset at that point.
+    run_podman --url $podman_socket pull $registry_flags $image_on_local_registry
+
+    run cat $_SOCAT_LOG
+    # The 'echo's help us debug failed runs
+    echo "socat log:"
+    echo "$output"
+    assert "$output" != ".*EXTEND_TIMEOUT_USEC=.*"
```
Comment (edsantiago) [inline test/system/260-sdnotify.bats:577]: one more suggestion:
```diff
-    assert "$output" != ".*EXTEND_TIMEOUT_USEC=.*"
+    assert "$output" !~ "EXTEND_TIMEOUT_USEC"
```

File: test/system/260-sdnotify.bats [hunk 5]
```diff
@@ -503,4 +505,79 @@ none | false | false | 0
...
+    wait_for_file $_SOCAT_LOG
+    local timeout=10
+    while [[ $timeout -gt 0 ]]; do
+        run cat $_SOCAT_LOG
+        # The 'echo's help us debug failed runs
+        echo "socat log:"
+        echo "$output"
+
+        if [[ "${lines[0]}" =~ "MAINPID=" ]]; then
+            break
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply test/system/260-sdnotify.bats:558]: Thanks for catching that!!

**Comment by vrothberg** [comment]
@edsantiago the APIv2 errors below look like flakes BUT `exist.io` actually exists, so I think the test should use localhost instead.

> [+0154s] not ok 1170 [60-auth] POST /v1.40/auth [-d {"username":"u9nBV48a","password":"WrOnGPassWord","serveraddress":"does.not.exist.io:1234/"}] : .message
[+0154s] #  expected: ~ .*no such host.*
[+0154s] #    actual: login attempt to does.not.exist.io:1234/ failed with status: authenticating creds for "does.not.exist.io:1234": pinging container registry does.not.exist.io:1234: Get "https://does.not.exist.io:1234/v2/": dial tcp: lookup does.not.exist.io on 169.254.169.254:53: server misbehaving

**Comment by edsantiago** [comment]
I.... think maybe not? As I read the test, it is important for it to fail with "no such host". `localhost` won't give us that. (For a moment I was tempted to suggest `localhsot` but even I am not that cruel).

Maybe `nxhost.nxdomain`? Or `nosuchhost.podman.io`? But even if we change that, I don't see it helping this particular failure: "server misbehaving" is new to me, but AFAICT it would not matter _what_ nonexistenthost we used in this test, it would still fail the same way.

**Comment by Luap99** [comment]
> `exist.io`

we can use something like `doesnotexists.podman.io` which is under our control

**Comment by vrothberg** [comment]
Rebased.  Ready for the final round of reviews.

**Comment by edsantiago** [comment]
LGTM minus a small typo that I'm not even going to point out because we need an excuse to run spellcheck once in a while.

@vrothberg why is gitbot auto-adding the WIP label?

**Comment by vrothberg** [comment]
> why is gitbot auto-adding the WIP label?

No idea. Maybe the api-chance implies WIP (would be strange)?

**Comment by vrothberg** [comment]
Should be good to merge :+1:

**Comment by Luap99** [comment]
> LGTM minus a small typo that I'm not even going to point out because we need an excuse to run spellcheck once in a while.
> 
> @vrothberg why is gitbot auto-adding the WIP label?

TOCTOU race conditions in the usage of the github API. I am not sure whose fault that is but I seen it before, it happens if two applications update labels at the same time. You can even trigger it via the webui.

**Review by Luap99 — APPROVED** [review]
/lgtm

**Pull request merged**